compiler/rustc_lint/src/traits.rs

```diff
@@ -132,3 +134,44 @@ impl<'tcx> LateLintPass<'tcx> for DropTraitConstraints {
         }
     }
 }
+
+declare_lint! {
+    pub DUP_TRAIT_BOUNDS,
+    Deny,
+    "Duplicate trait bounds are meaningless"
+}
+
+declare_lint_pass!(
+    DuplicateTraitBounds => [DUP_TRAIT_BOUNDS]
+);
+
+impl<'tcx> LateLintPass<'tcx> for DuplicateTraitBounds {
+    fn check_item(&mut self, cx: &LateContext<'_>, item: &'tcx hir::Item<'tcx>) {
+        let bounds: &[hir::GenericBound<'_>] = match &item.kind {
+            hir::ItemKind::Trait(_, _, _, bounds, _) => bounds,
+            hir::ItemKind::TraitAlias(_, bounds) => bounds,
+            _ => return,
+        };
+
+        let mut set = FxHashSet::default();
+        for bound in bounds.into_iter() {
+            match bound {
+                GenericBound::Trait(polytraitref, _) => {
+                    let Some(did) = polytraitref.trait_ref.trait_def_id() else { continue; };
+                    // If inserting the trait bound into the set returns `false`,
+                    // there is a duplicate.
+                    if !set.insert(did) {
```

Comment:
Instead of using a set, you should use a FxHashMap so that you can keep the Span around and use that Span in the lint to point at the previous trait bound, to minimize confusion.

Reply:
I'm not sure which methods to use here. Using this rough implementation

```
let mut map = FxHashMap::default();
        for bound in bounds.into_iter() {
            match bound {
                GenericBound::Trait(polytraitref, _) => {
                    let Some(did) = polytraitref.trait_ref.trait_def_id() else { continue; };
                    let span = polytraitref.span;
                    if map.insert(&did, &span).is_none() {
                        let span_of_dup_bound = span;
                        let span_of_prev_bound = map.get(&did).unwrap();
                        cx.struct_span_lint(DUP_TRAIT_BOUNDS, span_of_dup_bound, |lint| {
                            let msg = format!("duplicate trait bound");
                            lint.build(&msg)
                                .span_note(span_of_prev_bound, "trait bound previously used here")
                                .span_help(span_of_dup_bound, "remove this duplicate trait bound")
                                .emit();
                        });
                    };
                }
                _ => continue,
            }
        }
    }
```

I'm getting 

```
   --> compiler/rustc_lint/src/traits.rs:169:44
    |
169 | ...                   .span_help(span_of_prev_bound, "trait bound previously used here")
    |                        --------- ^^^^^^^^^^^^^^^^^^ the trait `From<&&rustc_span::Span>` is not implemented for `MultiSpan`
    |                        |
    |                        required by a bound introduced by this call
    |
    = help: the following other types implement trait `From<T>`:
              <MultiSpan as From<Vec<rustc_span::Span>>>
              <MultiSpan as From<rustc_span::Span>>
    = note: required because of the requirements on the impl of `Into<MultiSpan>` for `&&rustc_span::Span`
note: required by a bound in `DiagnosticBuilder::<'a, G>::span_help`
   --> /Users/will/repos/rust/compiler/rustc_errors/src/diagnostic_builder.rs:457:18
    |
457 |         sp: impl Into<MultiSpan>,
    |                  ^^^^^^^^^^^^^^^ required by this bound in `DiagnosticBuilder::<'a, G>::span_help`
```

I don't yet understand how to work with `MultiSpan`...

Also, for right now I'm thinking of this span message as a nice-to-have. I'm more worried about this test case and handling the generic parameters:

https://github.com/rust-lang/rust/blob/0b346be62c0dd278d52a6cc30cbb497a929cbbb1/src/test/ui/higher-rank-trait-bounds/normalize-under-binder/issue-81809.rs#L15

Reply:
That is because you have a `&&Span` and you need to dereference it to work.

Reply:
> That is because you have a &&Span and you need to dereference it to work.

Fun case, because then we should be suggesting the derefecence in the diagnostic, and we're not :)

Reply:
You can use [`try_insert`](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.try_insert):

```rust
let mut map = FxHashMap::default();
for bound in bounds.into_iter() {
    match bound {
        GenericBound::Trait(polytraitref, _) => {
            let Some(did) = polytraitref.trait_ref.trait_def_id() else { continue; };
            let span = polytraitref.span;
            if let Err(occupied) = map.try_insert(did, span) {
                cx.struct_span_lint(DUP_TRAIT_BOUNDS, span, |lint| {
                    let msg = format!("duplicate trait bound");
                    lint.build(&msg)
                        .span_label(*occupied.value, "trait bound previously used here")
                        .span_label(span, "duplicate trait bound")
                        .emit();
                });
            };
        }
        _ => continue,
    }
}
```